functions/AWSECSWrangleASG.js

```diff
@@ -0,0 +1,174 @@
+        if ( freeResourcesAfterWrangle > 0) {
+            console.log(`We will wrangle, as we have ${freeResourcesAfterWrangle} resource(s) available after this wrangle. This begins by draining the instance ${instanceWithMostResources.ID}`);
+            // We will wrangle, drain the instance in question
+            const drainingInstance = await ecsclient.send(
+                new UpdateContainerInstancesStateCommand({
+                    cluster: params.cluster,
+                    containerInstances: [].concat(instanceWithMostResources.ARN),
+                    status: "DRAINING"
+                }))
+
+            // Wait until the instance has 0 running tasks     
+            let drainingInstanceTimeoutCounter = 0;
+            let drainingInstanceRunningTasks = 1;
+            do {
+                // The division by 5 here is a throttling timeout pattern; to ensure that the ecsclient is not flooded with requests
+                await new Promise(x => setTimeout(x, (drainingInstanceTimeoutSeconds / 5) * 1000));
+                const drainingInstanceUpdate = await ecsclient.send(
+                    new DescribeContainerInstancesCommand({
+                        cluster: params.cluster,
+                        containerInstances: [].concat(instanceWithMostResources.ARN)
+                    }))
+
+                drainingInstanceRunningTasks = parseInt(drainingInstanceUpdate.containerInstances.runningTasksCount);
+
+                drainingInstanceTimeoutCounter++;
+
+                // Another part of throttling timeout pattern, 5 + 1 will trip the timeout and throw an error
+                if (drainingInstanceTimeoutCounter == 6) {
+                    throw `Tried to wrangle but instance did not drain within timeout of ${drainingInstanceTimeoutSeconds} seconds`
+                }
+            } while (drainingInstanceRunningTasks >> 0);
+
+            // Terminate the instance now that it has fully drained
+            const terminatingInstance = await asgclient.send(
+                new TerminateInstanceInAutoScalingGroupCommand({
+                    InstanceId: instanceWithMostResources.ID,
+                    ShouldDecrementDesiredCapacity: true
+                }))
+                
+            // Log the Auto Scaling Group activity that was triggered based off of the instance termination
+            console.log(`Terminated instance ${instanceWithMostResources.ID}, ASG replies with: ${terminatingInstance.Activity.Cause}`);
+            
+            // Sleep and run this script again, as we could have more instances to wrangle    
+            await new Promise(x => setTimeout(x, (reWrangleTimeoutSeconds * 1000)));
+            run();
+
+        } else {
+            // We will not wrangle
+            console.log(`We will not wrangle. This cluster's nodes are optimized`)
+        }
```

Comment:
This is more of a stylistic suggestion than anything else, but I personally have a strong preference for eliminating indents (i.e., keeping the "meat" of the function as close to the margin as possible) and preferring early exits/guard checks when possible. That would mea inverting this `if..else` so it would look more like this

```js
// Higher up, to prevent duplication
const exitPromise = new Promise(x => {
  setTimeout(() => x(), 1000)
});

// ... Then

if (freeResourcesAfterWrangle <= 0) {
  console.log(`We will not wrangle. This cluster's nodes are optimized`)
  return exitPromise;
}
```